Update protocol to support context integration with Cody

thenamankumar opened the pull request:
- Adds `mentions` feature to the Provider interface.
- Makes OpenCtx client automatically assign `provider` to every return type from the providers. 
- Adds `providerUri` filter arguments to the OpenCtx client methods so that the consumer can request capabilities, mentions and items from a particular provider. 
- Adds `meta` field to the `CapabilitiesResult`, providing the provider name. 
- Updates the CLI to support mentions, capabilities, providerUri arg and more.

Test plan

Same as: https://github.com/sourcegraph/openctx/pull/48

keegancsmith commented on client/vscode/src/api.ts:
```diff
@@ -10,15 +18,25 @@ export interface ExtensionApi {
+        /**
+         * Get OpenCtx provider capabilities.
+         */
+        capabilities(params: CapabilitiesParams): Promise<CapabilitiesResult[] | null>
```
@sqs we introduced this so that the vscode client can ask the provider what item types it provides. This is useful in the new autocomplete stuff around mentions where it lists all the types.

sqs replied:
makes sense

keegancsmith commented on client/vscode/src/api.ts:
```diff
@@ -10,15 +18,25 @@ export interface ExtensionApi {
          * Get OpenCtx items for the document.
          */
-        getItems(params: ItemsParams): Promise<Item[] | null>
+        items(params: ItemsParams): Promise<ItemsResult | null>
```
side-note: what is the motivation for all of these methods optionally returning null?

sqs replied:
no good reason. I guess `null` could mean "I knew I wouldn't have any results before I tried" and `[]` could mean "I might've had results but nah" but that is not clear or meaningful necessarily. So we can just remove a null option.

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -37,13 +40,17 @@ export interface ResponseError {
     data?: unknown
 }
-export interface CapabilitiesResult {
+export interface CapabilitiesResult<K extends ItemKindID = ItemKindID> {
```
Does the JSON schema generator understand this stuff? How does this sort of stuff impact the HTTP protocol?

sqs replied:
That file you edited is auto-generated from the JSON Schema. Don't make edits to this file directly.

thenamankumar replied:
Lol, we thought it was the other way around. I'll update the JSON schema. I don't expect it to putout any TS generics magic. @keegancsmith

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,11 +69,32 @@ export interface Selector {
+     * The list of respositoy remotes from the workspace.
+     * Eg. ["github.com/sourcegraph/openctx"]
+     */
+    repositoryRemotes?: string[]    
```
should we call this codebaseNames instead? The example above has been cleaned up and normalized, while I would expect "remote" to be more raw. From a provider implementor it seems rare to actually care about remote specifics / the client still needs to ensure it doesn't leak secrets from remote URLs.

sqs replied:
FWIW every time we have introduced an idea of a "repo name" or "codebase" into anything at sourcegraph, it has been imo a mistake. It is not a general concept and breaks when you consider monorepos, forks, weird git checkouts, non-git, etc. 

also typo `repositoy`

thenamankumar replied:
In that case we can probably skip this altogether. Right now it was required by the Github provider to get context from GH. But another way of doing is to ask user to manually map the workDir to git remote in the provider's settings.

I will skip adding ViewerContext for now.

keegancsmith replied:
@sqs do you have ideas around a better concept? I assume the idea of linking local state to a "shared state" (ie the repo you are contributing to) is a useful concept. Or is the issue auto-inference of what you are working on?

sqs replied:
I don't yet. It's really dependent on the specific situation. I mean, something like this is probably fine for now, but I think we should not hardcode it in the protocol spec and instead just pass it as a well-known key in some map.

There are a few ways the `repositoryRemote` would be used, and here are some common pitfalls:
- linking local repo to shared repo - need to ensure it works with forks, maybe should use repo ID (on Sourcegraph or GitHub) to not break when there are renames, some use cases should also use git SHAs to identify state
- understanding the user's current workspace - need to account for monorepo users who don't care about other subprojects or care less, need to account for multi-root workspace users
- other common pitfalls: normalizing git repo URLs is hard, often this breaks for github enterprise server etc, often you shoehorn other naming schemes like packages into a "repo name", often this breaks when the user has no git remotes set up yet (as @ErikaRS saw yesterday in her feedback)

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,11 +69,32 @@ export interface Selector {
+     */
+    query?: string
+    kind?: K
+    viewer?: Viewer
```
when we chatted you prefered adding this as an argument to the function. Did it end up cleaner to just use the Params field? TBH params seems like a reasonable place given the http layer.

sqs replied:
I think we should just keep this `{[key:string]: string}` for now and use a few well-known keys we agree on. Too soon to promote things into the protocol level yet before we've made a few more. For example I am not confident that repositoryRemotes or codebaseNames is the right abstraction (as mentioned above).

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,11 +69,32 @@ export interface Selector {
     /**
      * A search query that is interpreted by providers to filter the items in the result set.
      */
     query?: string
+    kind?: K
+    viewer?: Viewer
+    candidateItems?: CandidateItem[]
```
any chance we are making provider implementors lifes difficult by having such a wide param struct for both candidate and items function calls?

If I remember the original motivation for making it this big is we are thinking ahead to non-@mention cases where query can be a chat query?

keegancsmith commented on lib/provider/src/provider.ts:
```diff
@@ -1,31 +1,39 @@
      * Returns OpenCtx items.
      */
-    items?(params: ItemsParams, settings: S): ItemsResult | Promise<ItemsResult>
+    items?(params: ItemsParams<K>, settings: S): ItemsResult | Promise<ItemsResult>
```
it's a little unclear to me the value this generic adds. Can you motivate the benefits for my noobness.

thenamankumar replied:
While the generic will go away as the schema.ts will be generated from json and will provide no generics I think. 

The benefit of having a generic is that, while writing a provider in TS, we get a bit of type safety, as follows:

```typescript
type GithubItemKind = 'pr' | 'issue'
const MyProvider: <{}, 'pr' | 'issue'> = {
  items({kind}) {
    if (kind == 'pr') {
      return {
        kind: 'pr', // this will be typesafe and only allow either 'pr' or 'issue' strings. 
      }
    }
  }
}
```

But anyways, it will be removed.

thenamankumar replied:
Gone are the generics.

keegancsmith reviewed (commented):
nice, thanks! Will migrate my code on top of this :)

@sqs I left some inline comments about some motivations/etc. Would love your take on the API and your opinions/taste.

sqs commented on lib/schema/src/openctx.schema.ts:
```diff
@@ -1,18 +1,66 @@
  * Metadata about code
  */
-export type Schema = Item | Annotation
+export type Schema = CandidateItem | Item | Annotation
```
need to edit the `*.schema.json` files not these auto-generated files

thenamankumar replied:
Done.

sqs commented on lib/schema/src/openctx.schema.ts:
```diff
@@ -1,18 +1,66 @@
+/**
+ * A candidate item contains contextual information required to list and identify the resource.
+ */
+export interface CandidateItem<K extends ItemKindID = ItemKindID> {
```
Do we need to have CandidateItem be a separate type? It would be better for it to be an Item but with more leniency on fields being undefined. That would present a much simpler API surface imo. It would also make it easier to know if a candidate item is already present in the message, and to make other utility funcs to work with it (like to render it).

keegancsmith replied:
I like this suggestion a lot :+1:

thenamankumar replied:
Removed the separate type and using a single Item type only.

sqs commented on lib/schema/src/openctx.schema.ts:
```diff
@@ -1,18 +1,66 @@
+    /**
+     * A human-readable description of the item kind.
+     */
+    description?: string
```
What if 2 providers both return items of the same kind? I wonder if we should treat providers as a concept in the UI. Should we show the list of registered providers in the @-mention list, or should we show the kinds?

I think we should start with the former. That requires less "figuring out how to shove the provider author's semantic model of the world into consistency with that of other provider authors and us". They name their provider and then they're done. With item kinds, how do you deal with ItemKind.id conflicts? Should ItemKind.name be `Google Doc` or `Google Drive document` or `RFC`? It just requires an extra step of ascribing meaning that we don't know enough about yet as API designers.

thenamankumar replied:
Listing the providers from the client is a TODO and we definitely want to do that. But we will need a 2-level grouping. A single provider like Github, can support different item kinds like PRs & Issues, same with Jira or other 3rd party APIs. 

One way to render the list is:
```javascript
client.capabilities().flatMap(provider => provider.itemKinds.map(kind => 
  <div key={`${provider.uri}-${kind.id}`}>
    {kind.title} <sub>(by ${provider.title})</sub>
  </div>
))
```
```
-----------------------------
Github Issues (by Github)
-----------------------------
Github PRs (by Github)
-----------------------------
```

Second Option 
```javascript
client.capabilities().map(provider => 
  <div key={provider.uri} onClick={() => setSelected(provider)}>
    {provider.title}
  </div>
)

selected && selected.itemKinds.map(kind => <div key={kind.id}>{kind.title}</div>)
```
```
-----------------------------|-----------------------------
Github                       | Issues
-----------------------------|-----------------------------
                             | PRs
                             |-----------------------------
```

keegancsmith replied:
@sqs the main motivation behind this is I think it is reasonable for a provider to provide multiple types of context that a user may want to filter on. eg github provider does issues and PRs. A sourcegraph search provider could provide _lots_ of types (commits, packages, type:file, global (via dotcom), code intel?). I imagine a large customer would have lots of internal systems they may want to integrate with, and would probably deploy as a single provider?

sqs replied:
I see. And if we want to have a multiplexer here then we need one protocol counterparty to be able to represent stuff from multiple providers. Ie the provider is itself not a meaningful concept, there is no provider-specific implicit state.

This justifies a `kind` for me. So items would have `kind?: string` and capabilities would have `itemKinds: {id: string, title?: string}[]`? That makes sense to me.

I could see an argument for capabilities `itemKinds` being just `string[]` and being presentational, but the item kind `title` may need to be dynamic or we may want to have a separate icon/description field in the future. Neither of those are pressing needs, but they seem likely enough that we should make it an object with `{id: string, title?: string}` for now. It's annoying in a lot of languages to JSON-decode a `string | { id: string, ...}` value.

thenamankumar replied:
Thanks!

I think we are clear here. So resolving this thread.

sqs commented on lib/schema/src/openctx.schema.ts:
```diff
@@ -1,18 +1,66 @@
+/**
+ * Additional information which can be used to idenfity the resource by the provider.
+ */
+export type ItemMeta = { [key: string]: any }
```
What's an example of what this would contain? like the google docs document ID? or the github PR/issue number?

I think we should just rely on the URL for this. Providers can always encode extra stuff in a URL if needed.

I think using the URL suffices for all current known use cases. If we find this ItemMeta is truuuuly needed, we can add it later on.

keegancsmith replied:
I like the URL idea. My thinking around ItemMeta was there may be work done generating a candidate that you don't want to lose when generating the final item. However, that is indeed theoretical and we should only introduce when needed.

One potential issue with encoding stuff in the URL is we want it to remain clickable. Maybe we should rename URL to URI so we can encode extra stuff in it. Or I suppose if we wanted to encode lots of stuff into the uri you would use a custom url scheme.

thenamankumar replied:
I have removed the meta for now.

thenamankumar commented:
I have updated the right schema file, the json one and removed the separate CandidateItem type. 

The `ItemKind` is still there. Next, we will be working on exposing a list of providers from the OpenCtx client and updating its interface to support provider-type filters.

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -44,6 +47,20 @@ export interface CapabilitiesResult {
+     * Lists the kind of items the provider can return
+     */
+    itemKinds?: ItemKind[]
+    meta: {
```
why is this required?

thenamankumar replied:
If we need to list providers or even show that this context item is generated by this provider, we need the provider name. Provider URI isn't something we can show.

keegancsmith replied:
I guess this will all depend on how this information is shown to the user. @sqs has been working on the user facing stuff the most here so I respect his opinion the most here :)

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,11 +79,46 @@ export interface Selector {
+    /**
+     * The list of candidate items interpreted by providers to return items for the specified candidates.
+     */
+    candidateItems?: Item[]
```
I propose this just be a single item, rather than a list. Makes provider implementation more straightforward.

thenamankumar replied:
Done

keegancsmith commented on lib/ui-common/src/ui.test.ts:
```diff
@@ -11,7 +11,7 @@ describe('prepareAnnotationsForPresentation', () => {
                         end: { line: 5, character: 6 },
                     },
-                    item: { title: '1' },
+                    item: { title: '1', kind: 'file' },
```
kind is optional, did this need updating? I don't think kind plays a part in annotations right?

thenamankumar replied:
Yeah, I earlier kept it required. Removed the changes.

keegancsmith commented on lib/client/src/client/client.ts:
```diff
@@ -138,15 +189,15 @@ export interface Client<R extends Range> {
      * continues passing along any updates until unsubscribed.
      */
-    annotationsChanges(params: AnnotationsParams): Observable<Annotation<R>[]>
+    annotationsChanges(params: AnnotationsParams): Observable<EachWithProviderURI<Annotation<R>[]>>
```
The slapping everywhere of EachWithProviderURI makes me wonder if we need to adjust the protocol rather to encode this information.

A type like this makes me sweat.

thenamankumar replied:
What do you think is a better way to attach the provider URIs? I am open to ideas.

keegancsmith replied:
Maybe each field just has provider as part of the schema, but documented that the client overwrite this field and it doesn't need to be set by the provider? But yeah maybe this is the best way to do it from a TS point of view. I am still developing my TS taste :)

keegancsmith reviewed (commented):
Reviewed, but believe sqs is best to also review this.

sqs commented on lib/client/src/client/client.ts:
```diff
@@ -138,15 +197,21 @@ export interface Client<R extends Range> {
-    annotationsChanges(params: AnnotationsParams): Observable<Annotation<R>[]>
+    annotationsChanges(
+        params: AnnotationsParams,
+        providerUri?: string
```
What is the purpose of `providerUri`? Asking for myself now here and also you should add it in the docstring.

Why do you need to change the annotations API given that this work is about items?

thenamankumar replied:
The usage of `provider` is to get items/annotations for a specific provider. `item.kind` is expected to be unique only at the provider level, not globally. `providerUri` is the identifier for the provider. It value is based on the provider path the user sets in the OpenCtx config. 

We can skip making any changed to annotations for now.

keegancsmith replied:
Yeah lets follow-up. I believe its added here so you can route a request to a specific provider which maybe doesn't make sense for annotations, but is useful from a consistency in API point of view. So for example when passing in a mention we make sure to send it to the correct provider.

sqs commented on lib/client/src/client/client.ts:
```diff
@@ -112,6 +123,54 @@ export interface AuthInfo {
+     * The returned observable streams candidate items as they are received from the providers and continues
+     * passing along any updates until unsubscribed.
+     */
+    candidateItemsChanges(
```
Motivation:

1. I think we want a single API call for non-interactive applications that gets fully resolved items. Annoying if they have to call 2 APIs and make 2 roundtrips to do that. Can mean the server has discarded its internal state that would have made it just a tiny bit slower to get the fully resolved data, and instead the server needs to do 2x the number of API calls. We want to avoid that in our API design.
2. Many providers are just as fast returning candidate/preview items as they are the full items. Or the client will not need some of the omitted fields. It'd be nice for it to be a pure optimization, not a fundamental API concept, for the provider to return candidate/preview items.

My proposal:

1. For now, we just have `items` and no `candidateItems`. Yes, this means we will get back some data that might've been slow to fetch and is unnecessary to show in the @-mention menu. But it's simpler and probably fine for now.
2. When needed, as an optimization, then we can introduce a candidate items concept. Right now without seeing the use cases fully I am leaning toward channeling it all through the `items` method and letting you call `items` again with a set of possibly dehydrated/partial candidate items and asking it to fully hydrate them. In that way, the candidate items param to `items()` is like a query, so we can reuse an API concept there since we already have a query.

sqs commented on lib/client/src/client/client.ts:
```diff
@@ -112,6 +123,54 @@ export interface AuthInfo {
+     * RxJS), but it means that the client needs to manually poll for updated item kinds if freshness is
+     * important.
+     */
+    capabilities(
```
Not strictly related, but thoughts on renaming capabilities to meta? Capabilities was inspired by LSP but it's long and it's a confusing word, and we're calling the same thing meta elsewhere and that feels more natural.

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,12 +76,38 @@ export interface Selector {
+    /**
+     * A URI for the mention item.
+     */
+    uri?: string
```
in the slack thread I mentioned making this a requirement. but I guess given you added data it doesn't have to be required.

thenamankumar replied:
Updated, in the json schema, I made a typo, due to which it was left optional. Lets keep it required only.

thenamankumar commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,12 +76,38 @@ export interface Selector {
 }
+export interface ItemsParams {
+    /**
+     * A search query that is interpreted by providers to filter the items in the result set.
```
updated

keegancsmith commented on lib/client/src/client/client.ts:
```diff
@@ -112,6 +123,51 @@ export interface AuthInfo {
+     * RxJS), but it means that the client needs to manually poll for updated items if freshness is
+     * important.
+     */
+    mentions(params: MentionsParams, providerUri?: string): Promise<EachWithProviderUri<MentionsResult>>
```
ItemsParams doesn't contain a providerUri, but the response from this does. Is the idea that code that uses this client extracts out the providerURI from the mention and then sets it in the arguments to `items`.

thenamankumar replied:
The implementation of the client handles that. Neither the Provider nor the consumer of the client needs to manage that.

The providers are owned by Client, so it can assign the Uris.

keegancsmith commented on lib/client/src/client/client.ts:
```diff
@@ -236,31 +299,104 @@ export function createClient<R extends Range>(env: ClientEnv<R>): Client<R> {
+        )
+    }
+
+    // TODO: add option to get items from a specific provider.
```
Are these TODOs outdated? It looks like all the code around here respects the input providerUri

keegancsmith commented on lib/client/src/providerClient/createProviderClient.ts:
```diff
@@ -43,6 +53,28 @@ export function createProviderClient(
+            try {
+                return await transport.mentions(params, settings)
+            } catch (error) {
+                logger?.(`failed to get candidate items: ${error}`)
```
```suggestion
                logger?.(`failed to get mentions: ${error}`)
```

keegancsmith commented on lib/client/src/providerClient/createProviderClient.ts:
```diff
@@ -43,6 +53,28 @@ export function createProviderClient(
+            try {
+                return await transport.capabilities(params, settings)
+            } catch (error) {
+                logger?.(`failed to get item kinds: ${error}`)
```
```suggestion
                logger?.(`failed to get capabilities: ${error}`)
```

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,12 +76,38 @@ export interface Selector {
 }
+export interface ItemsParams {
+    /**
+     * A message that is interpreted by providers to return relevant items.
```
Just to check, in practice for our cody integration we will set both fields? IE message will be the final chat input and mention will what was selected from the mention dropdown?

I wonder if we should be clearer on the docstring here. For example in all our @ mention providers (except packages) we don't actually read this field and just rely on mention.

```suggestion
     * A message that can be interpreted by providers to return relevant items.
```

keegancsmith commented on provider/google-docs/index.ts:
```diff
@@ -38,7 +38,7 @@ const googleDocs: Provider<Settings> = {
         const driveAPI = googleDriveAPI({ version: 'v3', auth: oauth2Client })
 
-        const quotedQuery = JSON.stringify(params.query)
+        const quotedQuery = JSON.stringify(params.message)
```
As a TODO we need to update this provider to match the new API. If I am not mistaken this probably works better when run as a mention.

Also reading this provider makes me think message may be useful as part of mention as well. eg the search query below could do an or between the query or the chat message.

keegancsmith commented on provider/url-fetcher/index.ts:
```diff
@@ -16,21 +18,41 @@ const urlFetcher: Provider<UrlFetcherSettings> = {
+            return []
+        }
+
+        return [{ title: item.title, uri: item.url, data: { content: item.ai?.content } }]
```
data \o/

keegancsmith commented on lib/protocol/src/openctx-protocol.schema.ts:
```diff
@@ -62,12 +76,38 @@ export interface Selector {
+     */
+    uri: string
+    data?: {
+        [k: string]: unknown | undefined
```
is there a better type here for the value? We want to make sure that whatever the type here is it is json serializable.

thenamankumar replied:
This is generated code. Will figure out how we can update the JSON schema to generate a better type here later.

keegancsmith reviewed (approved):
LGTM! Some inline comments but lets land this beast.

keegancsmith commented:
🎉
